runtime/compiler/ilgen/Walker.cpp

```diff
@@ -3578,13 +3578,53 @@ TR_J9ByteCodeIlGenerator::genIfTwoOperand(TR::ILOpCodes nodeop)
 int32_t
 TR_J9ByteCodeIlGenerator::genIfImpl(TR::ILOpCodes nodeop)
    {
-   _methodSymbol->setHasBranches(true);
    int32_t branchBC = _bcIndex + next2BytesSigned();
    int32_t fallThruBC = _bcIndex + 3;
 
    TR::Node * second = pop();
    TR::Node * first = pop();
 
+   TR::DataType type = first->getDataType();
+   if (branchBC > _bcIndex &&
+       first->getOpCode().isLoadConst() &&
+       second->getOpCode().isLoadConst() &&
+       type != TR::Address &&
+       type != TR::Float &&
+       type != TR::Double)
+      {
+      int64_t v1 = first->getConstValue();
+      int64_t v2 = second->getConstValue();
+      bool branchTaken;
+      TR_ComparisonTypes compareType = TR::ILOpCode::getCompareType(nodeop);
+      switch (compareType)
+         {
+         case TR_cmpEQ:
+            branchTaken = v1 == v2;
+            break;
+         case TR_cmpNE:
+            branchTaken = v1 != v2;
+            break;
+         case TR_cmpLT:
+            branchTaken = v1 < v2;
+            break;
+         case TR_cmpLE:
+            branchTaken = v1 <= v2;
+            break;
+         case TR_cmpGT:
+            branchTaken = v1 > v2;
+            break;
+         case TR_cmpGE:
+            branchTaken = v1 >= v2;
+            break;
+         }
+
+      if (comp()->getOption(TR_TraceILGen))
+         traceMsg(comp(), "%s\n", branchTaken ? "taking the branch" : "fall through");
+
+      return branchTaken ? branchBC : fallThruBC;
```

Comment:
what about the _blocksToInline stuff below and associated gens - do we need to do that still?

Reply:
`_blocksToInline` is used in partial inlining, we still need it. What do you mean by associated gens?

Reply:
You are returning earlier than the follow-on code - if partial inlining is on or some other situation do we need to run the code below this point - it seems surprising that it is an unconditional return early.

Reply:
The partial inlining code seems complicated to deal with, I don't know how it supposed to work and I have never seen partial inlining kick in. I added a condition to only fold the if if we're not in partial inlining.

Reply:
where is this check?

Reply:
It is at line 3589.

Reply:
Wouldn't it be better to move that check down near the return and log a failure message about not folding due to partial inlining so it is easier to find this later if we need to fix it?

Reply:
Moved the check down